src/Discord/Internal/Types/ApplicationCommands.hs

```diff
@@ -0,0 +1,382 @@
+            ]
+      ]
+
+instance FromJSON ApplicationCommandOption where
+  parseJSON =
+    withObject
+      "ApplicationCommandOption"
+      ( \v ->
+          ApplicationCommandOption
+            <$> v .: "type"
+            <*> v .: "name"
+            <*> v .: "description"
+            <*> v .:? "required"
+            <*> v .:? "choices"
+            <*> v .:? "options"
+            <*> v .:? "channel_types"
+            <*> v .:? "min_val"
+            <*> v .:? "max_val"
+            <*> v .:? "autocomplete"
+      )
+
+-- | What type of command option. Can represent a wide variety of types, so
+-- please check out the documentation below.
+--
+-- https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-option-type
+data ApplicationCommandOptionType
+  = -- | A subcommand. It can take further options, excluding sub commands and
+    -- sub command groups.
+    ApplicationCommandOptionTypeSubCommand
+  | -- | A subcommand group. It can take further options, excluding sub command
+    -- groups.
+    ApplicationCommandOptionTypeSubCommandGroup
+  | -- | Can typically be provided with default values.
+    ApplicationCommandOptionTypeString
+  | -- | Can typically be provided with default values, and possibly with
+    -- minimum and maximum values.
+    ApplicationCommandOptionTypeInteger
+  | ApplicationCommandOptionTypeBoolean
+  | ApplicationCommandOptionTypeUser
+  | -- | Can be limited in the types of the channel allowed.
+    ApplicationCommandOptionTypeChannel
+  | ApplicationCommandOptionTypeRole
+  | -- | Users and roles.
+    ApplicationCommandOptionTypeMentionable
+  | -- | Can typically be provided with default values, and possibly with
+    -- minimum and maximum values. Represents a double.
+    ApplicationCommandOptionTypeNumber
+  deriving (Show, Read, Data, Eq)
+
+instance Enum ApplicationCommandOptionType where
+  fromEnum ApplicationCommandOptionTypeSubCommand = 1
+  fromEnum ApplicationCommandOptionTypeSubCommandGroup = 2
+  fromEnum ApplicationCommandOptionTypeString = 3
+  fromEnum ApplicationCommandOptionTypeInteger = 4
+  fromEnum ApplicationCommandOptionTypeBoolean = 5
+  fromEnum ApplicationCommandOptionTypeUser = 6
+  fromEnum ApplicationCommandOptionTypeChannel = 7
+  fromEnum ApplicationCommandOptionTypeRole = 8
+  fromEnum ApplicationCommandOptionTypeMentionable = 9
+  fromEnum ApplicationCommandOptionTypeNumber = 10
+  toEnum a = fromJust $ lookup a table
+    where
+      table = makeTable ApplicationCommandOptionTypeSubCommand
+
+instance ToJSON ApplicationCommandOptionType where
+  toJSON = toJSON . fromEnum
+
+instance FromJSON ApplicationCommandOptionType where
+  parseJSON = withScientific "ApplicationCommandOptionType" (return . toEnum . round)
+
+-- | Utility data type to store strings or number types.
+data StringNumberValue = StringNumberValueString String | StringNumberValueNumber Scientific
+  deriving (Show, Read, Eq)
```

Comment:
Looks nicer than Int/Double situation! 

String->Text for values. I think it's ok to do `StringNumberValueString T.Text` because text is a string type. Open to `TextNumberValueText Text` or `Either Text Scientific`

Reply:
I think the name should still be string to reflect the JSON descriptions of such things